### ISSUE TITLE
Uplift to Prometheus 3.x

### ISSUE THREAD

**Issue by jkuester** [medic/cht-watchdog#133]
Prometheus just [released `3.0`](https://prometheus.io/blog/2024/11/14/prometheus-3-0/#breaking-changes).  

> The Prometheus community strives to [not break existing features within a major release](https://prometheus.io/docs/prometheus/latest/stability/). With a new major release we took the opportunity to clean up a few, but small, long-standing issues. In other words, Prometheus 3.0 contains a few breaking changes. This includes changes to feature flags, configuration files, PromQL, and scrape protocols.

So, hopefully this will not require any big uplift, but we should look through the [migration guide](https://prometheus.io/docs/prometheus/3.0/migration/) to see if there is anything of concern...

### Good first issue

before anything, read the  [migration guide](https://prometheus.io/docs/prometheus/3.0/migration/) - make sure you understand the basics of the changes and how you might audit watchdog for being impacted.  The likely possible impact is the "Scrape protocols":

> This is a breaking change as scrapes that may have succeeded with Prometheus v2 may now fail if this fallback protocol is not specified.

Keep an eye out for this failing when working this ticket!

then, go ahead and make the change

1. start off by setting up an instance of CHT Core with [Docker Helper](https://docs.communityhealthtoolkit.org/hosting/4.x/app-developer/) 
2. spin up a local dev [instance](https://docs.communityhealthtoolkit.org/hosting/monitoring/setup/#setup) of watchdog by checking out the [repo](https://github.com/medic/cht-watchdog),  updating  the `cht-instances.yml` file to have the URL of your local dev CHT instance and running `docker compose up -d`
3. log into both CHT Core and watchdog - make sure the watchdog dashboards work as expected
4. edit the[ compose file](https://github.com/medic/cht-watchdog/blob/032ad6caa67518f8a9052cca3e8fcdd4d23805dd/docker-compose.yml#L18) to bump to latest Prometheus 3.x version (`3.2.1` as of [this writing](https://github.com/prometheus/prometheus/releases))
5. restart the watchdog containers with `docker compose restart`
6. make sure everything still works, that prometheus was indeed upgraded by looking at "build" on http://localhost:9090/status
7. check the logs with `docker compose logs -f` to see if there's any errors

**Comment by swet40** [comment]
@jkuester Can you assign this issue to me

**Comment by Omswastik-11** [comment]
Hi @jkuester !!

I am Om, I want to contribute to this issue. I have downloaded the cht-repo locally . I have experience in python . I am new to docker. Can I work on this ?

**Comment by jkuester** [comment]
👍 There will definitely be a learning curve here if you have never used Docker before.  But, feel free to give it a go!

**Comment by Omswastik-11** [comment]
Thanks .. I will try my best.

**Comment by Omswastik-11** [comment]
Hi @jkuester !!

I explored the Docker , Docker compose , and little bit about Prometheus and Graphana. I observed that latest Prometheus version is 3.3.0 but wrote to upgrade it to 3.2.1 ... should I try 3.3.0 ?

**Comment by Omswastik-11** [comment]
Hi @jkuester !!

I need some help while setting up CHT instances . I have few doubts these steps

1. " Edit the cht-instances.yml file to have the URLs of your CHT instances. You may include as many URLs of CHT instances as you like. " which links should I add

2. How I can get Watchdog dashboard ?


I have completed upto "docker compose up -d " It seems containers are running but no dashboard is shown . 

Do I need to login to Graphana ?

**Comment by Omswastik-11** [comment]
Hii @jkuester 

Why it showing error even after i give correct password and username but it is not able to sign in .. what should i do next ?

**Comment by jkuester** [comment]
>  I observed that latest Prometheus version is 3.3.0 but wrote to upgrade it to 3.2.1 ... should I try 3.3.0 

Thanks for checking that.  Yes, lets target `3.3.0`.  👍 

> Why grafana is showing error even after i give correct password and username but it is not able to sign in .. what should i do next ?

🤔 Did you follow the deployment steps [documented here](https://docs.communityhealthtoolkit.org/hosting/monitoring/setup/#setup)?  Can you see the Grafana login page at `http://localhost:3000/`?  You _should_ be able to login to Grafana out of the box with the user `medic` and the password that is set in your `.env` file. (And this should work independent of Prometheus).

**Comment by Omswastik-11** [comment]
Yes I have tried almost same steps. I manually edited the `.env` file added my own password saved and the runned the docker compose command. I opened the grafana from docker desktop . I given the password which I stored in ` .env` file. still it shows **Invalid user name or password**

`GRAFANA_ADMIN_USER=medic
GRAFANA_ADMIN_PASSWORD=Omswastik@1234`

This I editted manually in `.env` file

**Comment by mrjones-plip** [comment]
@Omswastik-11 - you might also consider fully deleting your watchdog install and starting from scratch.  Be aware this will [permanently delete all watchdog data](http://localhost:3000/)!  However, it will hopefully help you correctly setup a working Prometheus and Grafana instance.  

This assumes you have carefully followed the [setup steps](https://docs.communityhealthtoolkit.org/hosting/monitoring/setup/#setup) but can not log in:

1. go into your watchdog directory: `cd cht-watchdog`
2. kill and delete all watchdog containers, delete all networks and volumes: `docker compose kill;docker compose rm --force; docker compose down --volumes`
5. remove any on disk data for services: `sudo rm -rf  grafana/data  prometheus/data`
6. make sure your two `.ini` files are up to date: `cp grafana/grafana.example.ini grafana/grafana.ini`
7. recreate the data directories: `mkdir -p grafana/data && mkdir  -p prometheus/data`
8. startup the services in the background (detached): `docker compose up --detach`
9. get the username and password to login with: `egrep -i 'user|pass' .env | cut -f2
 -d=`
10. open http://localhost:3000/ in a browser and login with the username and password shown on prior step


Here's a video that shows this process:

https://github.com/user-attachments/assets/082b462a-4262-436a-8279-ff3fe1ffcaea

**Comment by Omswastik-11** [comment]
Thanks @mrjones-plip and @jkuester !!

It worked . I can now successfully login to grafana. 


but it seems there is no data for the grapfana to visualize. 

I want to ask 1 more question in compose file what changes I need to make ? I made `prom/prometheus` to pull the latest prometheus image. Is there some where else that I need to change some where ?

` prometheus:
    image: prom/prometheus
    ports:
      - "${PROMETHEUS_BIND:-127.0.0.1}:${PROMETHEUS_PORT:-9090}:9090"
    command:
      - '--config.file=/etc/prometheus/prometheus.yml'
      - '--storage.tsdb.path=/prometheus'
      - '--storage.tsdb.retention.time=${PROMETHEUS_RETENTION_TIME:-60d}'
      - '--web.console.libraries=/usr/share/prometheus/console_libraries'
      - '--web.console.templates=/usr/share/prometheus/consoles'
    volumes:
      - ./prometheus/config/prometheus.yml:/etc/prometheus/prometheus.yml:ro
      - ./cht-instances.yml:/etc/prometheus/cht-instances.yml:ro
      - ./exporters/json/config/scrape_config.yml:/etc/prometheus/scrape_configs/cht-json.yml:ro
      - prometheus-data:/prometheus
    logging:
      driver: "local"
      options:
        max-size: "${LOG_MAX_SIZE:-50m}"
        max-file: "${LOG_MAX_FILES:-20}"
    networks:
      - cht-watchdog-net
    restart: always`

**Comment by jkuester** [comment]
Glad to hear you got it working!

> there is no data for the grapfana to visualize. What should I do next ?

To easily test a local Watchdog deployment, you should:

- Run `docker compose down` to stop your current Watchdog instance.
- Apply the [development patches](https://github.com/medic/cht-watchdog/tree/main/development#development-patches). This just makes the data refresh in Prometheus/Grafana much faster than it does in production.
- Follow the [Stream Random Data](https://github.com/medic/cht-watchdog/blob/main/development/patches/apply/exporters.json.config.scrape_config.yml.patch#stream-random-data) instructions to run Watchdog along side a "fake" CHT instance that will report mock data.  

Ideally you should do this with the `main` branch of cht-watchdog first and confirm that you are seeing data pulling through as expected.  Spend some time clicking through the various Grafana dashboards and see the data being populated.  Then, once you have recorded 5-10 minutes of data, update the Prometheus version in the docker-compose and re-run:

```
docker compose -f docker-compose.yml -f exporters/postgres/compose.yml -f development/fake-cht/docker-compose.fake-cht.yml up -d
```

Again take some time to browse through Grafana and verify the various dashboards are still showing new data streaming in.  If anything was recording data pre-upgrade and then stopped showing data post-upgrade that is a problem and we will need to dig into what broke.

Once you are done testing and want to destroy your test instance along with the metric data collected, just run:

```
docker compose -f docker-compose.yml -f exporters/postgres/compose.yml -f development/fake-cht/docker-compose.fake-cht.yml down -v
```

**Comment by jkuester** [comment]
Great! Thanks for digging into this.  👍 Sounds like you should be good to raise a PR that updates the docker-compose image line to point to the latest Prometheus version.

> One thing there seems a warning

I think this is just Docker complaining that the [development compose file](https://github.com/medic/cht-watchdog/blob/032ad6caa67518f8a9052cca3e8fcdd4d23805dd/development/docker-compose.test-data.yml#L1) still has the old deprecated `version` property.  Feel free to just remove this property in the same PR. It should have no effect on the functionality (besides not printing the warning).

**Comment by Omswastik-11** [comment]
Ok, I will update the compose file to avoid the warning.

**Comment by Omswastik-11** [comment]
Hi @jkuester !! 
I raised a PR please review it if there is need of any changes .
